test/helpers/position_assertions.cc

```diff
@@ -1218,6 +1223,162 @@ string ApplyCompletionAssertion::toString() const {
+    : RangeAssertion(filename, range, assertionLine), version(string(version)), newName(newName), invalid(invalid),
+      expectedErrorMessage(expectedErrorMessage) {}
+
+void ApplyRenameAssertion::checkAll(const vector<shared_ptr<RangeAssertion>> &assertions,
+                                    const UnorderedMap<string, shared_ptr<core::File>> &sourceFileContents,
+                                    LSPWrapper &wrapper, int &nextId, string errorPrefix) {
+    for (auto assertion : assertions) {
+        if (auto assertionOfType = dynamic_pointer_cast<ApplyRenameAssertion>(assertion)) {
+            assertionOfType->check(sourceFileContents, wrapper, nextId, errorPrefix);
+        }
+    }
+}
+
+void ApplyRenameAssertion::check(const UnorderedMap<std::string, std::shared_ptr<core::File>> &sourceFileContents,
+                                 LSPWrapper &wrapper, int &nextId, std::string errorPrefix) {
+    auto prepareRenameResponse = doTextDocumentPrepareRename(wrapper, *this->range, nextId, this->filename);
+    // TODO: test placeholder in prepareRenameResponse
+
+    // A rename at an invalid position
+    if (newName.empty() && invalid) {
+        REQUIRE_EQ(prepareRenameResponse, nullptr);
+    } else {
+        REQUIRE_NE(prepareRenameResponse, nullptr);
+    }
+
+    auto workspaceEdits =
+        doTextDocumentRename(wrapper, *this->range, nextId, this->filename, newName, expectedErrorMessage);
+    // A rename at a valid position but with an invalid new name
+    if (invalid) {
+        REQUIRE_EQ(workspaceEdits, nullptr);
+        return;
+    }
+
+    {
+        INFO("doTextDocumentRename failed; see error above.");
+        REQUIRE_NE(workspaceEdits, nullptr);
+    }
+
+    auto &renameItems = *workspaceEdits->documentChanges;
+    auto it = sourceFileContents.find(this->filename);
+    {
+        INFO(fmt::format("Unable to find source file `{}`", this->filename));
+        REQUIRE_NE(it, sourceFileContents.end());
+    }
+
+    size_t index = this->filename.rfind("/", this->filename.length());
+    auto testDataPath = this->filename.substr(0, index);
+    // map of all .rb and .rbi files
+    UnorderedMap<string, string> sourceFiles;
+    // map of all .rbedited files whose version equals `this->version`
+    UnorderedMap<string, string> expectedEditedFiles;
+    for (auto filePath : FileOps::listFilesInDir(testDataPath, {".rb", ".rbi", ".rbedited"}, false, {}, {})) {
+        auto extension = FileOps::getExtension(filePath);
+        if (extension == "rbedited") {
+            auto extensionIndex = filePath.rfind(".rbedited", filePath.length());
+            auto fileVersion = string(FileOps::getExtension(filePath.substr(0, extensionIndex)));
+            if (fileVersion == this->version) {
+                expectedEditedFiles[filePath] = FileOps::read(filePath);
+            }
+        } else {
+            sourceFiles[filePath] = FileOps::read(filePath);
+        }
+    }
+
+    string sourceFilePath;
```

Comment:
Minor nit: These seem to only be used in the `for` loop; is there any reason why they are defined out here rather than at their first definition?